enable-font-awesome-lockdown-mode

vitorroriz opened the pull request:
#### 6198c51976a38d44659e41ca4791de9f763de19a
<pre>
enable-font-awesome-lockdown-mode
<a href="https://bugs.webkit.org/show_bug.cgi?id=248938">https://bugs.webkit.org/show_bug.cgi?id=248938</a>
rdar://101523138

Reviewed by Brent Fulgham.

* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/Ahem-10000A.ttf: Added.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/LockdownModeFonts.html: Added.
Adding new files: AllowedFonts.cpp and AllowedFonts.h

* Source/WebCore/css/CSSFontFace.cpp:
(WebCore::CSSFontFace::appendSources):
Adjusting allow downloading for modified settings &apos;DownloadableBinaryFontAllowedTypes&apos;.

* Source/WebCore/css/FontFace.cpp:
(WebCore::FontFace::create):
Adjusting allow downloading for modified settings &apos;DownloadableBinaryFontAllowedTypes&apos;.

* Source/WebCore/loader/cache/AllowedFonts.cpp: Added.
(WebCore::allowedFontHashesInLockdownMode):
(WebCore::hashForFontData):
(WebCore::isFontBinaryAllowed):
Source/WebCore/loader/cache/AllowedFonts.h: Added.
Contains definition for new enum class used by settings &apos;DownloadableBinaryFontAllowedTypes&apos;.
Contain hashes for the hard-coded allowed fonts when in restrict mode DownloadableBinaryFontAllowedTypes::Restricted.

* Source/WebCore/loader/cache/CachedFont.cpp:
(WebCore::CachedFont::shouldAllowCustomFont):
(WebCore::CachedFont::finishLoading):
(WebCore::CachedFont::setErrorAndDeleteData):
* Source/WebCore/loader/cache/CachedFont.h:
When font binary is finish loading, we will verify if we should allow custom font depending on the settings DownloadableBinaryFontAllowedTypes.
For DownloadableBinaryFontAllowedTypes::Any, we don&apos;t have to check if font can be loaded, any font is allowed.
For DownloadableBinaryFontAllowedTypes::Restricted, only fonts which binary-hash is listed in AllowedFonts.cpp:allowedHashes() will be loaded.
For DownloadableBinaryFontAllowedTypes::None, no font is downloaded or loaded.

When a font doesn&apos;t pass the loading test, we will not allow it to be loaded and its data buffer will get deleted.

* Source/WebCore/page/Quirks.cpp:
(WebCore::Quirks::shouldEnableFontLoadingAPIQuirk const):
Adjusting for modified settings &apos;DownloadableBinaryFontAllowedTypes&apos;.

* Source/WebCore/platform/SharedBuffer.h:
(WebCore::SharedBuffer::dataAsSpanForContiguousData const):
Adding function for returning data as Span if the buffer holds contiguous data. If data is not contiguous it returns a span with size 0.

* Source/WebCore/page/Settings.yaml:
* Source/WebCore/page/SettingsBase.h:
Modifying and renaming the settings &apos;DownloadableBinaryFontsEnabled&apos; to &apos;DownloadableBinaryFontAllowedTypes&apos;.
Now, instead of a bool, the settings is represented by a Enum with three values {Any, Restricted, None). See AllowedFonts.h.

* Source/WebKit/UIProcess/API/APIWebsitePolicies.cpp:
(API::WebsitePolicies::lockdownModeEnabled const):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::adjustSettingsForLockdownMode):
Adjusting for modified settings &apos;DownloadableBinaryFontAllowedTypes&apos;. Now, when in lockdown mode, instead of setting &apos;DownloadableBinaryFontsEnabled&apos; to true,
we set &apos;DownloadableBinaryFontAllowedTypes&apos; to &apos;Restricted&apos;. With that, font binaries will still be downloaded while in lockdown mode, but only allowed fonts will be loaded and kept in the system.

* Tools/TestWebKitAPI/Tests/WebKitCocoa/LockdownModeFonts.mm:
(TestWebKitAPI::TEST):
New test cases for loading allowed and not allowed fonts.
Modified FontLoadingApi and split it into two: AllowedFontLoadingApi and NotAllowedFontLoadingApi.
These will test that a allowed and not-allowed buffer can be loaded and lot loaded with the
FontFace font-loading API that utilizes data buffers.

Canonical link: <a href="https://commits.webkit.org/257954@main">https://commits.webkit.org/257954@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2906f204ac8ccffe3770e290950fa6b9cc28a9fe

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/100474 "Passed style check") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/77/builds/9633 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/43/builds/33535 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/8/builds/109808 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/170074 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/104465 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/10558 "Built successfully") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/85/builds/208 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/36/builds/92894 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/107661 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/106253 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/7989 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/43/builds/33535 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/92894 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/89902 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/43/builds/33535 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/36/builds/92894 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/81/builds/3368 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/43/builds/33535 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/82/builds/3361 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/85/builds/208 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/80/builds/9481 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/43/builds/33535 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5448 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/79/builds/5175 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

brentfulgham commented on Source/WebCore/loader/cache/CachedFont.cpp:
Propose adding:

```
bool CachedFont::shouldAllowCustomFont(const RefPtr<FragmentedSharedBuffer>& data)
{
    if (!m_loader || !m_loader->frame())
        return false;

    if (!m_loader->frame()->settings().shouldAllowLimitedUserInstalledFonts())
        return true;

    auto sha = hashForFontData(data->data(), data->size());
    return isAllowedFont(sha);
}

```

brentfulgham replied:
Create a new `WebPreferences.yaml` entry:

```
ShouldAllowLimitedUserInstalledFonts:
  type: bool
  webcoreOnChange: setNeedsRecalcStyleInAllFrames
  defaultValue:
    WebKitLegacy:
      default: false
    WebKit:
      default: false
    WebCore:
      default: false
```

brentfulgham commented on Source/WebCore/loader/cache/CachedFont.cpp:
I think the error method sets m_data to nullptr;

brentfulgham commented on Source/WebCore/loader/cache/CachedFont.cpp:
Then change this to:
```
if (! shouldAllowCustomFont(dataContiguous)) {
    ...
```

brentfulgham commented on Source/WebCore/loader/cache/CachedFont.cpp:
`m_data = WTFMove(dataContiguous);`

brentfulgham commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -4170,7 +4170,6 @@ static void adjustSettingsForLockdownMode(Settings& settings, const WebPreferenc
 #if ENABLE(WEB_AUDIO)
     settings.setWebAudioEnabled(false);
 #endif
-    settings.setDownloadableBinaryFontsEnabled(false);
```
Instead of deleting `settings.setDownloadableBinaryFontsEnabled`, change to:

`settings.setShouldAllowLimitedUserInstalledFonts(true);`

vitorroriz replied:
Should it be a setting (Settings.yaml) instead of a WebPreference (WebPreferences.yaml) ?

brentfulgham replied:
Oh! Yes, a Setting would be the right thing.

vitorroriz replied:
Should we simply replace one flag by the other and always allow to download but if the new flag is on we make the "allowed fonts check" , or you meant having 2 flags to describe different situations (DownloadableBinaryFontsEnabled, ShouldAllowLimitedUserInstalledFonts):
1. (false, false) : will not even try to download.
2. (false, true)  : will download, but just load the allowed fonts.
3. (true, false)  : will download and allow any fonts (skip the check for allowed fonts).
4. (true, true)   : should never happen, but if it does, the same as [3].

brentfulgham replied:
Maybe we should make it a tri-state value:

```
DownloadableBinaryFontAllowedTypes:
  type: uint32_t
  webcoreOnChange: setNeedsRecalcStyleInAllFrames
  defaultValue:
    WebKitLegacy:
      default: 2
    WebKit:
      default: 2
    WebCore:
      default: 2

```
Then you could make a "DownloadableBinaryFontAllowedTypes enum:

```
enum class DownloadableBinaryFontAllowedTypes : uint8_t {
    None = 0,
    Restricted = 1 << 0,
    Any = 1 << 1
};

```
'Any' would work like the current `setDownloadableBinaryFontsEnabled(true)`, 'None' would work like `setDownloadableBinaryFontsEnabled(false)`.

I don't think this is API exposed to clients, so I doubt there would be much compat risk.

brentfulgham commented on Source/WebCore/loader/cache/CachedFont.cpp:
Consider using `MemoryCompactLookupOnlyRobinHoodHashSet` once you have more than one entry.

brentfulgham commented on Source/WebCore/css/FontFace.cpp:
```diff
@@ -68,8 +69,8 @@ Ref<FontFace> FontFace::create(ScriptExecutionContext& context, const String& fa
-    bool allowDownloading = context.settingsValues().downloadableBinaryFontsEnabled;
-    auto sourceConversionResult = WTF::switchOn(source,
+    bool allowDownloading = context.settingsValues().downloadableBinaryFontAllowedTypes != DownloadableBinaryFontAllowedTypes::None;
+        auto sourceConversionResult = WTF::switchOn(source,
```
This shouldn't be indented.

brentfulgham commented on Source/WebCore/loader/cache/AllowedFonts.cpp:
```diff
@@ -0,0 +1,781 @@
+{
+    // Set of base64 strings of the sha256 representation of each allowed font binary.
+    static NeverDestroyed<MemoryCompactLookupOnlyRobinHoodHashSet<AtomString>> allowedFontHashes(std::initializer_list<AtomString> {
+    // fonts from app.slack.com
```
All of the font hashes should be indented.

brentfulgham reviewed (commented):
I think this looks good (modulo my whitespace nits), but I'd like to see if Chris has any suggestions for improvements.

I suspect we could use a more efficient lookup, and a smaller Hash structure than bin64 strings, but I think that could be a follow-up (as there are a number of places the same thing I done throughout WebKit).

cdumez commented on Source/WebCore/loader/cache/AllowedFonts.h:
```diff
@@ -0,0 +1,64 @@
+#include <wtf/EnumTraits.h>
+#include <wtf/RobinHoodHashSet.h>
+
+namespace WTF {
```
Should just include <wtf/Forward.h>

cdumez commented on Source/WebCore/loader/cache/AllowedFonts.h:
```diff
@@ -0,0 +1,64 @@
+
+namespace WebCore {
+
+const MemoryCompactLookupOnlyRobinHoodHashSet<WTF::AtomString>& allowedHashes();
```
Having a function with such a generic name `allowedHashes()` at the global WebCore level is not a good idea. Please use a more specific name, like `allowedFontHashes` or `allowedFontHashesInLockDownMode` or `allowedFontHashesInRestrictedMode`.

Also, you shouldn't need the `WTF::` before AtomString with my include suggestion above.

cdumez commented on Source/WebCore/loader/cache/CachedFont.cpp:
```diff
@@ -64,10 +68,47 @@ void CachedFont::didAddClient(CachedResourceClient& client)
+    auto cryptoDigest = PAL::CryptoDigest::create(PAL::CryptoDigest::Algorithm::SHA_256);
+    cryptoDigest->addBytes(data, size);
+    auto digest = cryptoDigest->computeHash();
+    return AtomString(makeString(base64Encoded(digest.data(), digest.size())));
```
Would `makeAtomString()` work instead of `AtomString(makeString())`. This would likely be more efficient if it did.

cdumez commented on Source/WebCore/loader/cache/CachedFont.cpp:
```diff
@@ -64,10 +68,47 @@ void CachedFont::didAddClient(CachedResourceClient& client)
         downcast<CachedFontClient>(client).fontLoaded(*this);
 }
 
+static AtomString hashForFontData(const uint8_t* data, size_t size)
```
We should take a `Span<const uint8_t>` in modern code instead of raw pointer + size.

cdumez commented on Source/WebCore/loader/cache/CachedFont.cpp:
```diff
@@ -64,10 +68,47 @@ void CachedFont::didAddClient(CachedResourceClient& client)
+    return AtomString(makeString(base64Encoded(digest.data(), digest.size())));
+}
+
+static bool isAllowedFont(const AtomString& fontSha256)
```
Not sure we need a utility function just for a single line doing a hash map look up.

cdumez commented on Source/WebCore/loader/cache/CachedFont.cpp:
```diff
@@ -64,10 +68,47 @@ void CachedFont::didAddClient(CachedResourceClient& client)
+static AtomString hashForFontData(const uint8_t* data, size_t size)
+{
+    auto cryptoDigest = PAL::CryptoDigest::create(PAL::CryptoDigest::Algorithm::SHA_256);
+    cryptoDigest->addBytes(data, size);
```
We may want to add an addBytes() overload that takes in a Span<const uint8_t>.

cdumez commented on Source/WebCore/loader/cache/CachedFont.cpp:
```diff
@@ -64,10 +68,47 @@ void CachedFont::didAddClient(CachedResourceClient& client)
+    case DownloadableBinaryFontAllowedTypes::None:
+        return false;
+    case DownloadableBinaryFontAllowedTypes::Restricted: {
+        auto sha = hashForFontData(data->data(), data->size());
```
We may want to add a `dataAsSpan()` function to SharedBuffer that returns a `Span<const uint8_t>`, and use it here.

cdumez commented on Source/WebCore/loader/cache/CachedFont.cpp:
```diff
@@ -77,6 +118,16 @@ void CachedFont::finishLoading(const FragmentedSharedBuffer* data, const Network
     checkNotify(metrics);
```
I am a little surprised we don't call setLoading(false) and checkNotify() in the new error case. Presumable we want to indicate that the load has finished, still.

vitorroriz replied:
`setErrorAndDeleteData()` calls `error(Status::DecodeError)` that will call both `setLoading(false)` and `checkNotify` with empty metrics.

brentfulgham commented on Source/WebCore/css/FontFace.cpp:
```diff
@@ -68,7 +69,7 @@ Ref<FontFace> FontFace::create(ScriptExecutionContext& context, const String& fa
     }
 
-    bool allowDownloading = context.settingsValues().downloadableBinaryFontsEnabled;
+    bool allowDownloading = context.settingsValues().downloadableBinaryFontAllowedTypes == DownloadableBinaryFontAllowedTypes::Any;
```
Should this be:

`context.settingsValues().downloadableBinaryFontAllowedTypes != DownloadableBinaryFontAllowedTypes::None;`

vitorroriz replied:
You are right, this is the place where we have to do the font hash check for the FontFace api case. I'll move the reusable code from CachedFont to AllowedFonts and add the check here.

brentfulgham reviewed (approved):
Thanks for creating a good test case for this! LGTM, but double-check that comparison I mentioned.

brentfulgham commented on Source/WebCore/loader/cache/AllowedFonts.h:
```diff
@@ -0,0 +1,62 @@
+};
+
+bool isFontBinaryAllowed(Span<const uint8_t>, DownloadableBinaryFontAllowedTypes);
+bool isFontBinaryAllowed(const void*, size_t, DownloadableBinaryFontAllowedTypes);
```
You only ever use the Span-based '`isFontBinaryAllowed`', so I'd remove the `void*` version.

brentfulgham commented on Source/WebCore/loader/cache/AllowedFonts.cpp:
```diff
@@ -0,0 +1,814 @@
+    case DownloadableBinaryFontAllowedTypes::None:
+        return false;
+    case DownloadableBinaryFontAllowedTypes::Restricted: {
+        auto sha = hashForFontData(data, size);
```
Just make this:

`auto sha = hashForFontData(data.data(), size.size());`

brentfulgham commented on Source/WebCore/loader/cache/AllowedFonts.cpp:
```diff
@@ -0,0 +1,814 @@
+    return makeAtomString(base64Encoded(digest.data(), digest.size()));
+}
+
+bool isFontBinaryAllowed(const void* data, size_t size, DownloadableBinaryFontAllowedTypes allowedType)
```
Just make this:

`bool isFontBinaryAllowed(Span<const uint8_t> data, DownloadableBinaryFontAllowedTypes allowedType)`

brentfulgham commented on Source/WebCore/loader/cache/AllowedFonts.cpp:
```diff
@@ -0,0 +1,814 @@
+    return false;
+}
+
+bool isFontBinaryAllowed(Span<const uint8_t> data, DownloadableBinaryFontAllowedTypes allowedType)
```
... and then remove this entirely.

brentfulgham commented on Source/WebCore/platform/SharedBuffer.h:
```diff
@@ -283,6 +283,7 @@ class SharedBuffer : public FragmentedSharedBuffer {
 
     WEBCORE_EXPORT const uint8_t* data() const;
     const char* dataAsCharPtr() const { return reinterpret_cast<const char*>(data()); }
+    Span<const uint8_t> dataAsSpanForContiguousData() const { return  Span(data(), isContiguous() ? size() : 0); }
```
Nit: Looks like two spaces between '`return`' and '`Span(...`'

brentfulgham commented on Source/WebCore/loader/cache/AllowedFonts.cpp:
```diff
@@ -0,0 +1,814 @@
+
+static AtomString hashForFontData(const void* data, size_t size)
+{
+    auto cryptoDigest = PAL::CryptoDigest::create(PAL::CryptoDigest::Algorithm::SHA_256);
```
Please add a comment above '`cryptoDigest->addBytes`':

`// FIXME(249414): Adopt Span in CryptoDigest`

brentfulgham reviewed (approved):
LGTM. I found a few minor changes I'd like, but no need for an additional review.

vitorroriz committed:
enable-font-awesome-lockdown-mode
https://bugs.webkit.org/show_bug.cgi?id=248938
rdar://101523138

Reviewed by Brent Fulgham.

* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/Ahem-10000A.ttf: Added.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/LockdownModeFonts.html: Added.
Adding new files: AllowedFonts.cpp and AllowedFonts.h

* Source/WebCore/css/CSSFontFace.cpp:
(WebCore::CSSFontFace::appendSources):
Adjusting allow downloading for modified settings 'DownloadableBinaryFontAllowedTypes'.

* Source/WebCore/css/FontFace.cpp:
(WebCore::FontFace::create):
Adjusting allow downloading for modified settings 'DownloadableBinaryFontAllowedTypes'.

* Source/WebCore/loader/cache/AllowedFonts.cpp: Added.
(WebCore::allowedFontHashesInLockdownMode):
(WebCore::hashForFontData):
(WebCore::isFontBinaryAllowed):
Source/WebCore/loader/cache/AllowedFonts.h: Added.
Contains definition for new enum class used by settings 'DownloadableBinaryFontAllowedTypes'.
Contain hashes for the hard-coded allowed fonts when in restrict mode DownloadableBinaryFontAllowedTypes::Restricted.

* Source/WebCore/loader/cache/CachedFont.cpp:
(WebCore::CachedFont::shouldAllowCustomFont):
(WebCore::CachedFont::finishLoading):
(WebCore::CachedFont::setErrorAndDeleteData):
* Source/WebCore/loader/cache/CachedFont.h:
When font binary is finish loading, we will verify if we should allow custom font depending on the settings DownloadableBinaryFontAllowedTypes.
For DownloadableBinaryFontAllowedTypes::Any, we don't have to check if font can be loaded, any font is allowed.
For DownloadableBinaryFontAllowedTypes::Restricted, only fonts which binary-hash is listed in AllowedFonts.cpp:allowedHashes() will be loaded.
For DownloadableBinaryFontAllowedTypes::None, no font is downloaded or loaded.

When a font doesn't pass the loading test, we will not allow it to be loaded and its data buffer will get deleted.

* Source/WebCore/page/Quirks.cpp:
(WebCore::Quirks::shouldEnableFontLoadingAPIQuirk const):
Adjusting for modified settings 'DownloadableBinaryFontAllowedTypes'.

* Source/WebCore/platform/SharedBuffer.h:
(WebCore::SharedBuffer::dataAsSpanForContiguousData const):
Adding function for returning data as Span if the buffer holds contiguous data. If data is not contiguous it returns a span with size 0.

* Source/WebCore/page/Settings.yaml:
* Source/WebCore/page/SettingsBase.h:
Modifying and renaming the settings 'DownloadableBinaryFontsEnabled' to 'DownloadableBinaryFontAllowedTypes'.
Now, instead of a bool, the settings is represented by a Enum with three values {Any, Restricted, None). See AllowedFonts.h.

* Source/WebKit/UIProcess/API/APIWebsitePolicies.cpp:
(API::WebsitePolicies::lockdownModeEnabled const):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::adjustSettingsForLockdownMode):
Adjusting for modified settings 'DownloadableBinaryFontAllowedTypes'. Now, when in lockdown mode, instead of setting 'DownloadableBinaryFontsEnabled' to true,
we set 'DownloadableBinaryFontAllowedTypes' to 'Restricted'. With that, font binaries will still be downloaded while in lockdown mode, but only allowed fonts will be loaded and kept in the system.

* Tools/TestWebKitAPI/Tests/WebKitCocoa/LockdownModeFonts.mm:
(TestWebKitAPI::TEST):
New test cases for loading allowed and not allowed fonts.
Modified FontLoadingApi and split it into two: AllowedFontLoadingApi and NotAllowedFontLoadingApi.
These will test that a allowed and not-allowed buffer can be loaded and lot loaded with the
FontFace font-loading API that utilizes data buffers.

Canonical link: https://commits.webkit.org/257954@main
